clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -494,10 +580,37 @@ private void handleDisconnections(List<ClientResponse> responses, long now) {
     /**
      * Record any newly completed connections
      */
-    private void handleConnections() {
+    private void handleConnections(long now) {
         for (String node : this.selector.connected()) {
             log.debug("Completed connection to node {}", node);
-            this.connectionStates.connected(node);
+            // Though the node is connected, we might not still be able to send requests. For instance,
+            // in case of SSL connection, SSL handshake happens after connection is established.
+            if (this.requiredApiVersions == null) {
+                this.connectionStates.ready(node);
+            } else {
+                this.connectionStates.checkingApiVersions(node);
+                if (!apiVersionsRequests.contains(node))
+                    apiVersionsRequests.add(node);
+            }
+        }
+        maybeInitiateApiVersionsFetch(now);
```

Comment:
Never mind. I talked with Jun and Ismael about this and "connected" currently does not imply "ready."